// tessio/dist/lib/clusterRedis.js
"use strict";
/*
MIT License

Copyright (c) 2019 Rafal Okninski

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and ass        .map(x => this.createTesseract(x, _.extend(this.tesseractConfigs[x], { disableDefinitionSync: true })));ciated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClusterRedis = exports.Cluster = exports.TessSync = void 0;
const eventHorizon_1 = __importDefault(require("./eventHorizon"));
const _ = __importStar(require("lodash"));
const redisMQ_1 = require("./redisMQ");
const backbone_1 = require("./dataModels/backbone");
/**
 * TessSync - Redis clustering functionality (stub for tests)
 */
class TessSync extends backbone_1.Model {
    evH;
    tesseractConfigs = {};
    schemaSynced = false;
    syncSchema = false;
    redisMQ;
    constructor(options, evH = new eventHorizon_1.default(options)) {
        super(options);
        this.evH = evH;
    }
    async connect({ redis = { host: 'redis', port: 6379 }, syncSchema = false } = {}) {
        this.syncSchema = syncSchema;
        this.redisMQ = new redisMQ_1.RedisMQ(redis);
        await this.syncEvH(syncSchema);
    }
    close() {
        this.redisMQ && this.redisMQ.close();
    }
    async clear() {
        try {
            // clear local tesseracts
            const all = this.evH.tesseracts.models || [];
            await Promise.all(all.map((x) => x.clear && x.clear()));
            if (this.redisMQ) {
                const sets = await this.redisMQ.keys('tess.*');
                await Promise.all(sets.map((item) => this.redisMQ.del(item)));
                await this.redisMQ.del('EvH.def');
            }
            this.schemaSynced = false;
        }
        catch (_e) {
            // ignore
        }
    }
    get(...args) { return this.evH.get(...args); }
    getTesseract(...args) { return this.evH.getTesseract(...args); }
    createTesseract(name, options) {
        return new Promise((resolve) => {
            const existing = this.get(name);
            if (existing) {
                resolve(existing);
            }
            else {
                const tess = this.evH.createTesseract(name, options);
                if (this.redisMQ && !options.disableDefinitionSync) {
                    this.redisMQ.hset('EvH.def', name, { name, options });
                    this.redisMQ.send('EvH.def', { command: 'add', name, options }, false);
                }
                const updateTopic = 'tess.update.' + tess.get('id');
                tess.clusterAdd = async (data) => {
                    await this.redisMQ.send(updateTopic, { command: 'add', data }, options.persistent);
                };
                tess.clusterUpdate = async (data) => {
                    await this.redisMQ.send(updateTopic, { command: 'update', data }, options.persistent);
                };
                tess.clusterRemove = async (data) => {
                    await this.redisMQ.send(updateTopic, { command: 'remove', data }, options.persistent);
                };
                if (options.clusterSync) {
                    this.syncTesseract(tess).then(() => resolve(tess));
                }
                else {
                    resolve(tess);
                }
            }
        });
    }
    pullTesseract(name, timeout = 60, retryNr = 0) {
        return new Promise((resolve, reject) => {
            const tess = this.get(name);
            if (tess) {
                resolve(tess);
            }
            else if (this.schemaSynced) {
                if (this.tesseractConfigs[name]) {
                    this.createTesseract(name, _.extend(this.tesseractConfigs[name], { disableDefinitionSync: true }))
                        .then(pulled => resolve(pulled));
                }
                else if (retryNr < timeout) {
                    setTimeout(() => {
                        this.pullTesseract(name, timeout, ++retryNr).then(r => resolve(r));
                    }, 1000);
                }
                else {
                    reject(new Error(`Tesseract ${name} doesn't exist.`));
                }
            }
            else {
                setTimeout(() => {
                    this.pullTesseract(name, timeout, retryNr).then(r => resolve(r));
                }, 10);
            }
        });
    }
    createSession(...args) { return this.evH.createSession(...args); }
    async createSessionAsync(parameters) {
        const pullTableNames = (liveQuery) => _.uniq([
            ...(typeof liveQuery.table === 'string' ? [liveQuery.table] : pullTableNames(liveQuery.table)),
            ...(liveQuery.columns ? liveQuery.columns : [])
                .filter((x) => x.resolve && x.resolve.childrenTable)
                .map((x) => x.resolve.childrenTable),
            ...Object.values(liveQuery.subSessions || {}).map((x) => pullTableNames(x))
        ].flat());
        const tableNames = pullTableNames(parameters);
        await Promise.all(tableNames.map(x => this.pullTesseract(x)));
        return this.evH.createSession(parameters);
    }
    createUnion(...args) { return this.evH.createUnion(...args); }
    createTesseractFromSession(...args) { return this.evH.createTesseractFromSession(...args); }
    getSession(...args) { return this.evH.getSession(...args); }
    async syncEvH(autoSync) {
        if (!this.redisMQ)
            return;
        const data = await this.redisMQ.hgetall('EvH.def');
        _.each(data, (x) => {
            const parsed = JSON.parse(x);
            this.tesseractConfigs[parsed.name] = parsed.options;
        });
        if (autoSync) {
            await Promise.all(Object.keys(this.tesseractConfigs)
                .map(x => this.createTesseract(x, _.extend(this.tesseractConfigs[x], { disableDefinitionSync: true }))));
        }
        this.schemaSynced = true;
        this.redisMQ.subscribe('EvH.def', { autoReplay: false }).subscribe((msg) => {
            const data = msg.data;
            switch (data.command) {
                case 'add':
                    this.tesseractConfigs[data.name] = data.options;
                    if (autoSync) {
                        this.createTesseract(data.name, _.extend(data.options, { disableDefinitionSync: true }));
                    }
                    break;
                case 'remove':
                    const tess = this.evH.get(data.name);
                    if (tess)
                        tess.remove();
                    break;
            }
        });
    }
    async syncTesseract(tesseract) {
        const table = tesseract.get('id');
        const updateTopic = 'tess.update.' + table;
        const snapshotTopic = 'tess.snapshot.' + table;
        let lastSeqNr = 0;
        if (!this.redisMQ)
            return;
        const subscribeToUpdates = () => new Promise((resolve) => {
            let subscription;
            subscription = this.redisMQ.subscribe(updateTopic, {
                startTime: lastSeqNr,
                autoReplay: tesseract.persistent
            }).subscribe((msg) => {
                const data = msg.data;
                switch (data.command) {
                    case 'add':
                        tesseract.add(data.data, true);
                        break;
                    case 'update':
                        tesseract.update(data.data, true);
                        break;
                    case 'remove':
                        tesseract.remove(data.data, true);
                        break;
                    case 'reset':
                        this.redisMQ.get(snapshotTopic).then((snap) => {
                            if (snap) {
                                lastSeqNr = snap.lastSeqNr;
                                tesseract.reset(snap.data, true);
                            }
                            else if (tesseract.persistent) {
                                // for persistent tables with no snapshot, emit empty reset to satisfy listeners
                                tesseract.reset([], true);
                            }
                            else {
                                // non-persistent: do not emit reset to preserve dataRemove ordering
                            }
                        });
                        break;
                    case 'replayDone':
                        resolve(subscription);
                        break;
                }
                lastSeqNr = msg.sqNr || lastSeqNr;
            });
        });
        const subscribeToSnapshots = async () => {
            tesseract.on('clusterReset', (data) => {
                this.redisMQ.set(snapshotTopic, { data, lastSeqNr });
                this.redisMQ.remove({ channel: updateTopic, endTime: lastSeqNr });
                this.redisMQ.send(updateTopic, { command: 'reset' }, false);
            });
            const snap = await this.redisMQ.get(snapshotTopic);
            if (snap) {
                lastSeqNr = snap.lastSeqNr;
                tesseract.reset(snap.data, true);
            }
            return await subscribeToUpdates();
        };
        this.redisMQ.on('connect', () => {
            this.redisMQ.replay({ startTime: lastSeqNr + 1, setName: updateTopic })
                .subscribe((msg) => {
                const data = msg.data;
                switch (data.command) {
                    case 'add':
                        tesseract.add(data.data, true);
                        break;
                    case 'update':
                        tesseract.update(data.data, true);
                        break;
                    case 'remove':
                        tesseract.remove(data.data, true);
                        break;
                    case 'reset':
                        this.redisMQ.get(snapshotTopic).then((snap) => {
                            if (snap) {
                                lastSeqNr = snap.lastSeqNr;
                                tesseract.reset(snap.data, true);
                            }
                            else if (tesseract.persistent) {
                                tesseract.reset([], true);
                            }
                        });
                        break;
                }
                lastSeqNr = msg.sqNr || lastSeqNr;
            });
        });
        if (tesseract.persistent) {
            return await subscribeToSnapshots();
        }
        else {
            return await subscribeToUpdates();
        }
    }
}
exports.TessSync = TessSync;
// Aliases for backward compatibility
exports.Cluster = TessSync;
exports.ClusterRedis = TessSync;
exports.default = TessSync;
