// tessio/dist/lib/redisMQ.js
"use strict";
/*
MIT License

Copyright (c) 2019 Rafal Okninski

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
 
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisMQ = void 0;
const ioredis_1 = __importDefault(require("ioredis"));
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const baseTime = BigInt(new Date().getTime() * 1000000000);
const basehrTime = process.hrtime.bigint();
class RedisMQ {
    pub;
    sub;
    constructor(redisConfig = {}) {
        this.pub = new ioredis_1.default(redisConfig);
        this.sub = new ioredis_1.default(redisConfig);
        this.sub.setMaxListeners(10000);
    }
    on(event, listener) {
        return this.sub.on(event, listener);
    }
    async remove({ channel, startTime = 0, endTime = '+inf' }) {
        return this.pub.zremrangebyscore(channel, startTime, endTime);
    }
    close() {
        this.pub.disconnect();
        this.sub.disconnect();
    }
    async del(channel) {
        return this.pub.del(channel);
    }
    async keys(pattern) {
        return this.pub.keys(pattern);
    }
    replay(options) {
        return new rxjs_1.Observable((observer) => {
            const startTime = options.startTime || 0;
            const endTime = options.endTime || '+inf';
            const batchSize = options.batchSize || 10000;
            let offset = 0;
            let exit = false;
            const processBatch = async (setName) => {
                const batch = await this.pub.zrangebyscore(setName, startTime, endTime, 'WITHSCORES', 'LIMIT', offset, batchSize);
                const batchLength = batch.length / 2;
                for (let i = 0; i < batch.length; i += 2) {
                    const message = JSON.parse(batch[i]);
                    observer.next(message);
                }
                return {
                    length: batch.length,
                    full: batchLength === batchSize
                };
            };
            const loop = async () => {
                const { full } = await processBatch(options.setName);
                if (full && !exit) {
                    offset += batchSize;
                    loop();
                }
                else {
                    observer.complete();
                }
            };
            loop();
            return () => {
                exit = true;
            };
        });
    }
    subscribe(topic, options = {}) {
        const { startTime = 0, autoReplay = true } = options;
        const subMessage$ = (0, rxjs_1.fromEvent)(this.sub, 'message');
        return new rxjs_1.Observable(observer => {
            const subscription = subMessage$
                .pipe((0, operators_1.filter)((data) => {
                const [channel] = data;
                return channel === topic;
            }))
                .subscribe(([, message]) => {
                const obj = JSON.parse(message);
                observer.next(obj);
            });
            let replaySub;
            if (autoReplay) {
                replaySub = this.replay({
                    setName: topic,
                    startTime
                }).subscribe(message => observer.next(message), error => observer.error(error), () => {
                    this.sub.subscribe(topic);
                    observer.next({
                        data: {
                            command: 'replayDone'
                        }
                    });
                });
            }
            else {
                this.sub.subscribe(topic);
                observer.next({
                    data: {
                        command: 'replayDone'
                    }
                });
            }
            return () => {
                if (autoReplay && replaySub) {
                    replaySub.unsubscribe();
                }
                subscription.unsubscribe();
            };
        });
    }
    async send(setName, data, persistant = true) {
        const obj = {
            data: data,
            sqNr: (baseTime + (process.hrtime.bigint() - basehrTime)).toString()
        };
        const message = JSON.stringify(obj);
        if (persistant) {
            await this.pub.zadd(setName, obj.sqNr, message);
        }
        await this.pub.publish(setName, message);
        return obj;
    }
    async get(setName) {
        const data = await this.pub.get(setName);
        return data ? JSON.parse(data) : null;
    }
    async set(setName, data) {
        return this.pub.set(setName, JSON.stringify(data));
    }
    async hget(setName, field) {
        const data = await this.pub.hget(setName, field);
        return data ? JSON.parse(data) : null;
    }
    async hgetall(setName) {
        return this.pub.hgetall(setName);
    }
    async hset(setName, field, data) {
        return this.pub.hset(setName, field, JSON.stringify(data));
    }
}
exports.RedisMQ = RedisMQ;
