// tessio/dist/lib/tesseract.js
"use strict";
/*
MIT License

Copyright (c) 2019 Rafal Okninski

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tesseract = void 0;
const backbone_1 = require("./dataModels/backbone");
const _ = __importStar(require("lodash"));
const linq = __importStar(require("linq"));
const utils_1 = require("./utils");
const expressionEngine_1 = require("./expressionEngine");
const UPDATE_REASON_DATA = 'dataUpdate';
const UPDATE_REASON_DATA_RESET = 'dataReset';
const expressionEngine = new expressionEngine_1.ExpressionEngine();
class Tesseract extends backbone_1.Model {
    id;
    columns;
    dataCache = [];
    dataMap = {};
    sessions;
    idProperty = 'id';
    idIndex = 0;
    clusterSync = false;
    persistent = false;
    secondaryIndexes = [];
    dataIndex = {};
    hasGarbage = false;
    objDef;
    defferedDataUpdateTime = 0;
    updatedRows = [];
    removedRows = [];
    defferedDataUpdate;
    defferedDataRemove;
    refreshData;
    refreshTesseract;
    collectGarbage;
    // Wire-in resolver function from constructor options (EventHorizon.resolve)
    resolverFn;
    constructor(config) {
        super(config);
        this.id = config.id || this.generateGuid();
        // Support both TS-style { id } and legacy JS-style { name }
        this.columns = (config.columns || []).map((raw) => {
            const name = (raw && (raw.name ?? raw.id));
            const col = {
                name,
                title: raw?.title,
                primaryKey: raw?.primaryKey,
                secondaryKey: !!raw?.secondaryKey,
                value: raw?.value,
                defaultValue: raw?.defaultValue,
                columnType: raw?.columnType,
                aggregator: raw?.aggregator,
                expression: raw?.expression,
                resolve: raw?.resolve,
                enum: raw?.enum,
            };
            return col;
        });
        this.clusterSync = !!config.clusterSync;
        this.persistent = !!config.persistent;
        this.defferedDataUpdateTime = config.defferedDataUpdateTime || 0;
        // Save resolver function from config
        this.resolverFn = config.resolve;
        // Find primary key
        const pkColumn = this.columns.find(col => col.primaryKey);
        if (pkColumn) {
            this.idProperty = pkColumn.name;
        }
        this.idIndex = this.columns.findIndex(col => col.name === this.idProperty);
        this.sessions = new backbone_1.Collection();
        this.objDef = this.generateObjDef(this.columns);
        if (config.data) {
            this.dataCache = this.generateData(config.data);
            this.generateIndex();
        }
        // Initialize debounced functions
        this.refreshData = _.debounce(() => {
            this.trigger('dataUpdate', this.dataCache);
        }, 100, { maxWait: 100 });
        this.refreshTesseract = _.debounce((silent) => {
            this.dataCache.forEach(row => {
                this.generateRow(row, this.columns, row);
            });
            if (!silent) {
                this.trigger('dataUpdate', this.dataCache);
            }
        }, 100, { maxWait: 100 });
        this.collectGarbage = _.debounce(() => {
            if (this.hasGarbage) {
                this.dataCache = this.dataCache.filter(x => !x.removed);
                this.hasGarbage = false;
            }
        }, 100, { maxWait: 100 });
        this.updatedRows = [];
        this.removedRows = [];
        if (this.defferedDataUpdateTime > 0) {
            this.defferedDataUpdate = _.debounce((updatedData, disableClusterUpdate, updateReason) => {
                this.trigger('dataUpdate', updatedData, disableClusterUpdate, updateReason);
                this.updatedRows = [];
            }, 10, { maxWait: this.defferedDataUpdateTime });
            this.defferedDataRemove = _.debounce((removedRows, disableClusterUpdate, updateReason) => {
                this.trigger('dataRemove', removedRows, disableClusterUpdate, updateReason);
                this.removedRows = [];
            }, 10, { maxWait: this.defferedDataUpdateTime });
        }
    }
    generateIndex() {
        this.secondaryIndexes = [];
        this.columns.forEach(column => {
            if (column.secondaryKey) {
                this.secondaryIndexes.push(column.name);
                this.dataIndex[column.name] = {};
                linq.from(this.dataCache)
                    .groupBy((x) => x[column.name])
                    .forEach((x) => {
                    this.dataIndex[column.name][x.key()] = x.getSource();
                });
            }
        });
    }
    updateIndex(newData, dataHolder) {
        if (this.secondaryIndexes.length === 0)
            return;
        const idProperty = this.idProperty;
        const orgData = this.dataMap[newData[idProperty]];
        this.secondaryIndexes.forEach(indexName => {
            if (!orgData) {
                if (!this.dataIndex[indexName][newData[indexName]]) {
                    this.dataIndex[indexName][newData[indexName]] = [];
                }
                this.dataIndex[indexName][newData[indexName]].push(dataHolder);
            }
            else if (newData[indexName] !== undefined && orgData[indexName] !== newData[indexName]) {
                if (!this.dataIndex[indexName][newData[indexName]]) {
                    this.dataIndex[indexName][newData[indexName]] = [];
                }
                this.dataIndex[indexName][newData[indexName]].push(dataHolder);
                const tempValue = orgData[indexName];
                const index = this.dataIndex[indexName][tempValue].findIndex(x => x[idProperty] === newData[idProperty]);
                this.dataIndex[indexName][tempValue].splice(index, 1);
            }
        });
        this.dataMap[newData[this.idProperty]] = newData;
    }
    removeFromIndex(newData) {
        delete this.dataMap[newData[this.idProperty]];
        if (this.secondaryIndexes.length === 0)
            return;
        const idProperty = this.idProperty;
        this.secondaryIndexes.forEach(indexName => {
            const tempValue = newData[indexName];
            const index = this.dataIndex[indexName][tempValue].findIndex(x => x[idProperty] === newData[idProperty]);
            this.dataIndex[indexName][tempValue].splice(index, 1);
        });
    }
    get(stuff) {
        if (this[stuff]) {
            return this[stuff];
        }
    }
    createSession(config) {
        const id = config.id || this.generateGuid();
        // Import Session dynamically to avoid circular dependency
        const Session = require('./session').Session;
        const session = new Session({
            id,
            tesseract: this,
            config,
            getTesseract: config.getTesseract || this.getTesseract.bind(this)
        });
        this.sessions.add(session);
        return session;
    }
    getTesseract() {
        return new Promise((resolve) => {
            resolve(this);
        });
    }
    getData() {
        if (this.hasGarbage) {
            this.dataCache = this.dataCache.filter(x => !x.removed);
            this.hasGarbage = false;
        }
        return this.dataCache;
    }
    getLinq() {
        let tempData = linq.from(this.dataCache);
        if (this.hasGarbage) {
            this.dataCache = tempData.where((x) => !x.removed).toArray();
            tempData = linq.from(this.dataCache);
            this.hasGarbage = false;
        }
        return tempData;
    }
    getCount() {
        if (this.hasGarbage) {
            this.dataCache = this.dataCache.filter(x => !x.removed);
            this.hasGarbage = false;
        }
        return this.dataCache.length;
    }
    getById(id) {
        return this.dataMap[id];
    }
    async add(data, disableClusterUpdate = false) {
        if (!data)
            return [];
        if (!Array.isArray(data)) {
            data = [data];
        }
        const tempRows = [];
        if (this.clusterSync && !disableClusterUpdate) {
            await this.clusterAdd(data);
        }
        else {
            for (let i = 0; i < data.length; i++) {
                let tempRow = this.dataMap[data[i][this.idProperty]];
                if (!tempRow) {
                    tempRow = this.generateRow(data[i], this.columns);
                    this.dataCache.push(tempRow);
                    this.updatedRows.push(tempRow);
                    tempRows.push(tempRow);
                }
            }
            if (this.updatedRows.length > 0) {
                if (this.defferedDataUpdateTime > 0) {
                    if (this.removedRows.length !== 0 && this.defferedDataRemove) {
                        this.defferedDataRemove.flush();
                    }
                    if (this.defferedDataUpdate) {
                        this.defferedDataUpdate(this.updatedRows, disableClusterUpdate, UPDATE_REASON_DATA);
                    }
                }
                else {
                    this.trigger('dataUpdate', this.updatedRows, disableClusterUpdate, UPDATE_REASON_DATA);
                    this.updatedRows = [];
                }
            }
        }
        return tempRows;
    }
    addAsync(data, disableClusterUpdate = false) {
        return new Promise((resolve, reject) => {
            if (!data) {
                reject();
                return;
            }
            if (!Array.isArray(data)) {
                data = [data];
            }
            const handler = (x) => {
                if (x && x.length && x[0][this.idProperty] === data[0][this.idProperty]) {
                    this.off('dataUpdate', handler);
                    resolve(x);
                }
            };
            this.on('dataUpdate', handler);
            this.add(data, disableClusterUpdate);
        });
    }
    reset(data, disableClusterUpdate = false, suppressEvents = false) {
        if (this.clusterSync && !disableClusterUpdate) {
            this.trigger('clusterReset', data || this.getData());
        }
        else if (data) {
            this.dataMap = {};
            this.dataCache = this.generateData(data);
            if (!suppressEvents) {
                if (this.defferedDataUpdateTime > 0 && this.defferedDataUpdate) {
                    this.defferedDataUpdate.flush();
                }
                this.trigger('dataUpdate', this.dataCache, true, UPDATE_REASON_DATA_RESET);
            }
            else {
                this.clearSessions();
            }
        }
        return this.dataCache;
    }
    async update(data, disableClusterUpdate = false) {
        if (!data)
            return [];
        if (!Array.isArray(data)) {
            data = [data];
        }
        if (data.length === 0)
            return [];
        const tempRows = [];
        if (this.clusterSync && !disableClusterUpdate) {
            await this.clusterUpdate(data);
        }
        else {
            for (let i = 0; i < data.length; i++) {
                let tempRow;
                if (Array.isArray(data[i])) {
                    tempRow = this.dataMap[data[i][this.idIndex]];
                }
                else {
                    tempRow = this.dataMap[data[i][this.idProperty]];
                }
                if (tempRow) {
                    this.generateRow(data[i], this.columns, tempRow);
                    this.updatedRows.push(tempRow);
                    tempRows.push(tempRow);
                }
                else {
                    tempRow = this.generateRow(data[i], this.columns);
                    this.updatedRows.push(tempRow);
                    this.dataCache.push(tempRow);
                    tempRows.push(tempRow);
                }
            }
            if (this.updatedRows.length > 0) {
                if (this.defferedDataUpdateTime > 0) {
                    if (this.removedRows.length !== 0 && this.defferedDataRemove) {
                        this.defferedDataRemove.flush();
                    }
                    if (this.defferedDataUpdate) {
                        this.defferedDataUpdate(this.updatedRows, disableClusterUpdate, UPDATE_REASON_DATA);
                    }
                }
                else {
                    this.trigger('dataUpdate', this.updatedRows, disableClusterUpdate, UPDATE_REASON_DATA);
                    this.updatedRows = [];
                }
            }
        }
        return tempRows;
    }
    updateAsync(data, disableClusterUpdate = false) {
        return new Promise((resolve, reject) => {
            if (!data) {
                reject();
                return;
            }
            if (!Array.isArray(data)) {
                data = [data];
            }
            const handler = (x) => {
                if (x && x.length && x[0][this.idProperty] === data[0][this.idProperty]) {
                    this.off('dataUpdate', handler);
                    resolve(x);
                }
            };
            this.on('dataUpdate', handler);
            this.update(data, disableClusterUpdate);
        });
    }
    async remove(data, disableClusterUpdate = false) {
        if (!Array.isArray(data))
            data = [data];
        if (this.clusterSync && !disableClusterUpdate) {
            await this.clusterRemove(data);
        }
        else {
            for (let i = 0; i < data.length; i++) {
                const tempRow = this.dataMap[data[i]];
                if (tempRow) {
                    tempRow.removed = true;
                    this.removedRows.push(tempRow[this.idProperty]);
                    this.hasGarbage = true;
                    this.removeFromIndex(tempRow);
                }
            }
            if (this.removedRows.length > 0) {
                if (this.defferedDataUpdateTime > 0) {
                    if (this.defferedDataUpdate && this.defferedDataUpdate.flush) {
                        this.defferedDataUpdate.flush();
                    }
                    if (this.defferedDataRemove) {
                        this.defferedDataRemove(this.removedRows, disableClusterUpdate);
                    }
                }
                else {
                    this.trigger('dataRemove', this.removedRows, disableClusterUpdate);
                    this.removedRows = [];
                }
            }
        }
    }
    removeAsync(data, disableClusterUpdate = false) {
        return new Promise((resolve, reject) => {
            if (data === undefined || data === null) {
                reject();
                return;
            }
            const ids = Array.isArray(data) ? data : [data];
            const handler = (removedIds) => {
                if (Array.isArray(removedIds) && ids.some(id => removedIds.includes(id))) {
                    this.off('dataRemove', handler);
                    resolve(removedIds);
                }
            };
            this.on('dataRemove', handler);
            this.remove(ids, disableClusterUpdate);
        });
    }
    clear(disableClusterUpdate, suppressEvents = false) {
        // Handle backbone-style options object vs tesseract-style boolean parameters
        if (typeof disableClusterUpdate === 'object') {
            // Backbone-style call - delegate to parent
            return super.clear(disableClusterUpdate);
        }
        // Tesseract-style call
        const disableCluster = disableClusterUpdate || false;
        return new Promise((resolve) => {
            const ids = this.dataCache.map(x => x[this.idProperty]);
            if (this.clusterSync && !disableCluster) {
                const done = () => {
                    this.off('dataRemove', onRemove);
                    this.off('dataUpdate', onUpdate);
                    resolve();
                };
                const onRemove = () => done();
                const onUpdate = (_data, _disable, reason) => {
                    if (reason === UPDATE_REASON_DATA_RESET)
                        done();
                };
                this.on('dataRemove', onRemove);
                this.on('dataUpdate', onUpdate);
                // Ensure remove is processed before reset to preserve event order
                const send = async () => {
                    try {
                        await this.clusterRemove(ids);
                    }
                    catch { /* ignore */ }
                    this.trigger('clusterReset', []);
                };
                send();
            }
            else {
                if (!suppressEvents) {
                    this.trigger('dataRemove', ids);
                }
                else {
                    // Clear sessions quietly
                    if (this.sessions && this.sessions.models) {
                        this.sessions.models.forEach((x) => x.clear && x.clear());
                    }
                }
                this.dataCache = [];
                this.dataMap = {};
                resolve();
            }
        });
    }
    updateColumns(newColumns, reset) {
        let updatedColumns = reset ? newColumns : newColumns;
        this.columns = updatedColumns;
        this.objDef = this.generateObjDef(this.columns);
        this.refresh();
    }
    refresh() {
        if (this.refreshTesseract) {
            this.refreshTesseract(false);
        }
    }
    generateData(data) {
        if (!data)
            return [];
        return data.map(row => this.generateRow(row, this.columns));
    }
    generateRow(data, columns, dataHolder, updateDataMap = true) {
        let copyData = true;
        let arrayDataType = false;
        if (dataHolder !== undefined) {
            updateDataMap = false;
            if (dataHolder === data) {
                copyData = false;
            }
        }
        else if (data instanceof this.objDef) {
            dataHolder = data;
        }
        else {
            dataHolder = new this.objDef();
        }
        this.updateIndex(data, dataHolder);
        if (Array.isArray(data)) {
            arrayDataType = true;
        }
        for (let i = 0; i < columns.length; i++) {
            const column = columns[i];
            const propertyName = column.name;
            const defaultValue = column.defaultValue;
            const value = column.value;
            const propertyValue = arrayDataType ? data[i] : data[propertyName];
            if (defaultValue !== undefined && dataHolder[propertyName] === undefined) {
                const valueType = typeof defaultValue;
                if (valueType === 'function') {
                    dataHolder[propertyName] = defaultValue(dataHolder, propertyValue, propertyName);
                }
                else if (valueType === 'string') {
                    dataHolder[propertyName] = _.template.call(dataHolder, defaultValue)(dataHolder);
                }
                else {
                    dataHolder[propertyName] = defaultValue;
                }
            }
            if (value !== undefined) {
                const valueType = typeof value;
                if (valueType === 'function') {
                    dataHolder[propertyName] = value(dataHolder, propertyValue, dataHolder[propertyName], propertyName);
                }
                else if (valueType === 'string') {
                    dataHolder[propertyName] = _.template.call(dataHolder, value)(dataHolder);
                }
                else {
                    dataHolder[propertyName] = value;
                }
            }
            else if (copyData && propertyValue !== undefined) {
                dataHolder[propertyName] = propertyValue;
            }
            if (column.expression !== undefined) {
                if (column.expressionTree === undefined) {
                    column.expressionTree = expressionEngine.generateExpressionTree(column.expression);
                    console.log('Generated expression tree for', column.name, column.expressionTree);
                }
                if (column.expressionTree) {
                    dataHolder[propertyName] = expressionEngine.executeExpressionTree(column.expressionTree, dataHolder);
                    console.log('Executed expression for', column.name, 'result:', dataHolder[propertyName]);
                }
            }
            if (column.resolve !== undefined && this.resolverFn) {
                dataHolder[propertyName] = this.resolverFn(column.resolve, dataHolder);
            }
        }
        if (updateDataMap) {
            this.dataMap[dataHolder[this.idProperty]] = dataHolder;
        }
        return dataHolder;
    }
    generateObjDef(columns) {
        let classMeat = 'return class {';
        columns.forEach((column) => {
            classMeat += `"${column.name}"\n`;
        });
        classMeat += '"removed"\n}';
        // eslint-disable-next-line no-new-func
        return (new Function(classMeat))();
    }
    returnTree(rootIdValue, parentIdField, groups) {
        const root = this.dataMap[rootIdValue];
        if (root) {
            if (!groups) {
                groups = _.groupBy(this.dataCache, (x) => x[parentIdField]);
            }
            const newItem = root;
            if (groups[newItem[this.idProperty]] && groups[newItem[this.idProperty]].length) {
                newItem.children = [];
                groups[newItem[this.idProperty]].forEach((x) => {
                    if (x[this.idProperty] !== x[parentIdField]) {
                        const childrenItem = this.returnTree(x[this.idProperty], parentIdField, groups);
                        newItem.children.push(childrenItem);
                    }
                });
            }
            newItem.leaf = !newItem.children;
            return newItem;
        }
    }
    getHeader(excludeHiddenColumns) {
        return (0, utils_1.getHeader)(this.columns, excludeHiddenColumns);
    }
    getSimpleHeader(excludeHiddenColumns) {
        return (0, utils_1.getSimpleHeader)(this.columns, excludeHiddenColumns);
    }
    // Cluster methods placeholders
    async clusterAdd(_data) {
        // Implementation for cluster operations would go here
    }
    async clusterUpdate(_data) {
        // Implementation for cluster operations would go here
    }
    async clusterRemove(_data) {
        // Implementation for cluster operations would go here
    }
    destroy() {
        this.trigger('destroy');
        this.off();
    }
    generateGuid() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            const r = Math.random() * 16 | 0;
            const v = c === 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
    clearSessions() {
        try {
            // Try to access sessions and clear them
            const sessions = this.sessions;
            if (sessions && sessions.models && Array.isArray(sessions.models)) {
                sessions.models.forEach((x) => x.clearSession && x.clearSession());
            }
            else if (sessions && sessions.forEach) {
                sessions.forEach((x) => x.clearSession && x.clearSession());
            }
        }
        catch (e) {
            // Ignore errors in clearing sessions
        }
    }
}
exports.Tesseract = Tesseract;
exports.default = Tesseract;
