// orm3/dist/SQLQuery/Dialects/mssql.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const helpers = __importStar(require("../Helpers"));
const DataTypes = {
    id: 'INT IDENTITY(1,1) NOT NULL PRIMARY KEY',
    int: 'INT',
    float: 'FLOAT',
    bool: 'BIT',
    text: 'TEXT'
};
function escapeId(...identifiers) {
    return identifiers
        .map((identifier) => {
        if (typeof identifier === 'object' && identifier !== null) {
            return String(identifier.str).replace(/\?:(id|value)/g, (match) => {
                if (match === '?:id') {
                    return escapeId.apply(this, [identifier.escapes.shift()]);
                }
                return escapeVal.call(this, identifier.escapes.shift());
            });
        }
        return `[${identifier}]`;
    })
        .join('.');
}
function escapeVal(val, timeZone) {
    if (val === undefined || val === null || typeof val === 'symbol') {
        return 'NULL';
    }
    if (Array.isArray(val)) {
        if (val.length === 1 && Array.isArray(val[0])) {
            return `(${val[0].map((item) => escapeVal.call(this, item, timeZone)).join(',')})`;
        }
        return `(${val.map((item) => escapeVal.call(this, item, timeZone)).join(', ')})`;
    }
    if (val instanceof Date) {
        return `'${helpers.dateToString(val, timeZone || 'local', { dialect: 'mssql' })}'`;
    }
    if (Buffer.isBuffer(val)) {
        return `X'${val.toString('hex')}'`;
    }
    switch (typeof val) {
        case 'number':
            if (!Number.isFinite(val)) {
                return `'${val.toString()}'`;
            }
            return String(val);
        case 'boolean':
            return val ? '1' : '0';
        case 'function':
            return val(mssqlDialect);
        case 'string':
            break;
        default:
            return `'${JSON.stringify(val).replace(/'/g, "''")}'`;
    }
    return `'${val.replace(/'/g, "''")}'`;
}
function escape(query, args) {
    return helpers.escapeQuery(mssqlDialect, query, args);
}
const mssqlDialect = {
    DataTypes,
    escape,
    escapeId: (...ids) => escapeId.apply(mssqlDialect, ids),
    escapeVal: (value, timeZone) => escapeVal.call(mssqlDialect, value, timeZone),
    defaultValuesStmt: 'DEFAULT VALUES',
    limitAsTop: true
};
exports.default = mssqlDialect;
