// orm3/dist/SQLQuery/Dialects/mysql.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const helpers = __importStar(require("../Helpers"));
const DataTypes = {
    id: 'INTEGER PRIMARY KEY AUTO_INCREMENT',
    int: 'INTEGER',
    float: 'FLOAT(12,2)',
    bool: 'TINYINT(1)',
    text: 'TEXT'
};
function escapeId(...identifiers) {
    return identifiers
        .map((identifier) => {
        if (typeof identifier === 'object' && identifier !== null) {
            return String(identifier.str).replace(/\?:(id|value)/g, (match) => {
                if (match === '?:id') {
                    return escapeId.apply(this, [identifier.escapes.shift()]);
                }
                return escapeVal.call(this, identifier.escapes.shift());
            });
        }
        return `\`${String(identifier).replace(/`/g, '``')}\``;
    })
        .join('.');
}
function escapeVal(val, timeZone) {
    if (val === undefined || val === null || typeof val === 'symbol') {
        return 'NULL';
    }
    if (Buffer.isBuffer(val)) {
        return bufferToString(val);
    }
    if (Array.isArray(val)) {
        return arrayToList.call(this, val, timeZone || 'local');
    }
    let value = val;
    if (value instanceof Date) {
        value = helpers.dateToString(value, timeZone || 'local', { dialect: 'mysql' });
    }
    else {
        switch (typeof value) {
            case 'boolean':
                return value ? 'true' : 'false';
            case 'number':
                if (!Number.isFinite(value)) {
                    value = value.toString();
                    break;
                }
                return String(value);
            case 'object':
                return objectToValues.call(this, value, timeZone || 'Z');
            case 'function':
                return value(mysqlDialect);
            default:
                break;
        }
    }
    const escaped = String(value).replace(/[\0\n\r\b\t\\\'"\x1a]/g, (s) => {
        switch (s) {
            case '\0':
                return '\\0';
            case '\n':
                return '\\n';
            case '\r':
                return '\\r';
            case '\b':
                return '\\b';
            case '\t':
                return '\\t';
            case "'":
                return "\\'";
            case '"':
                return '\\"';
            case '\x1a':
                return '\\Z';
            case '\\':
                return '\\\\';
            default:
                return s;
        }
    });
    return `'${escaped}'`;
}
function objectToValues(object, timeZone) {
    const values = [];
    for (const key in object) {
        if (!Object.prototype.hasOwnProperty.call(object, key)) {
            continue;
        }
        const value = object[key];
        if (typeof value === 'function') {
            continue;
        }
        values.push(`${escapeId.call(this, key)} = ${escapeVal.call(this, value, timeZone)}`);
    }
    return values.join(', ');
}
function arrayToList(array, timeZone) {
    return `(${array
        .map((value) => {
        if (Array.isArray(value)) {
            return arrayToList.call(this, value, timeZone);
        }
        return escapeVal.call(this, value, timeZone);
    })
        .join(', ')})`;
}
function bufferToString(buffer) {
    let hex;
    try {
        hex = buffer.toString('hex');
    }
    catch (err) {
        const chunks = [];
        for (let i = 0; i < buffer.length; i += 1) {
            chunks.push(zeroPad(buffer[i].toString(16)));
        }
        hex = chunks.join('');
    }
    return `X'${hex}'`;
}
function zeroPad(value) {
    return value.length === 1 ? `0${value}` : value;
}
function escape(query, args) {
    return helpers.escapeQuery(mysqlDialect, query, args);
}
const mysqlDialect = {
    DataTypes,
    escape,
    escapeId: (...ids) => escapeId.apply(mysqlDialect, ids),
    escapeVal: (value, timeZone) => escapeVal.call(mysqlDialect, value, timeZone),
    defaultValuesStmt: 'VALUES()'
};
exports.default = mysqlDialect;
