// orm3/dist/SQLQuery/Dialects/sqlite.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers = __importStar(require("../Helpers"));
const mysql_1 = __importDefault(require("./mysql"));
const DataTypes = {
    isSQLITE: true,
    id: 'INTEGER PRIMARY KEY AUTOINCREMENT',
    int: 'INTEGER',
    float: 'FLOAT(12,2)',
    bool: 'TINYINT(1)',
    text: 'TEXT'
};
function escapeVal(val, timeZone) {
    if (val === undefined || val === null || typeof val === 'symbol') {
        return 'NULL';
    }
    if (Array.isArray(val)) {
        if (val.length === 1 && Array.isArray(val[0])) {
            return `(${val[0].map((item) => escapeVal.call(this, item, timeZone)).join(',')})`;
        }
        return `(${val.map((item) => escapeVal.call(this, item, timeZone)).join(', ')})`;
    }
    if (val instanceof Date) {
        return `'${helpers.dateToString(val, timeZone || 'local', { dialect: 'sqlite' })}'`;
    }
    if (Buffer.isBuffer(val)) {
        return `X'${val.toString('hex')}'`;
    }
    switch (typeof val) {
        case 'number':
            if (!Number.isFinite(val)) {
                return `'${val.toString()}'`;
            }
            return String(val);
        case 'boolean':
            return val ? '1' : '0';
        case 'function':
            return val(sqliteDialect);
        case 'string':
            break;
        default:
            return `'${JSON.stringify(val).replace(/'/g, "''")}'`;
    }
    return `'${val.replace(/'/g, "''")}'`;
}
function escape(query, args) {
    return helpers.escapeQuery(sqliteDialect, query, args);
}
const sqliteDialect = {
    DataTypes,
    escape,
    escapeId: (...ids) => mysql_1.default.escapeId(...ids),
    escapeVal: (value, timeZone) => escapeVal.call(sqliteDialect, value, timeZone),
    defaultValuesStmt: 'DEFAULT VALUES'
};
exports.default = sqliteDialect;
