// orm3/dist/Validators/Validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Validator {
    constructor(validate) {
        this.validator = validate;
    }
    validate(value, next, thisArg, contexts = {}) {
        this.validator.apply(thisArg, [value, next, contexts]);
    }
    ifDefined() {
        return new Validator((value, next, contexts) => {
            if (value === undefined || value === null)
                return next();
            return this.validator(value, next, contexts);
        });
    }
    ifNotEmptyString() {
        return new Validator((value, next, contexts) => {
            if (value === undefined || value === null)
                return next();
            if (typeof value !== 'string')
                return next();
            if (value.length === 0)
                return next();
            return this.validator(value, next, contexts);
        });
    }
    ifType(type) {
        return new Validator((value, next, contexts) => {
            if (typeof value !== type)
                return next();
            return this.validator(value, next, contexts);
        });
    }
    ifNotType(type) {
        return new Validator((value, next, contexts) => {
            if (typeof value !== type)
                return this.validator(value, next, contexts);
            return next();
        });
    }
}
exports.default = Validator;
