// orm3/dist/Validators/Validators.js
"use strict";
/**
 * Validators module for ORM
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validators = void 0;
exports.equalToProperty = equalToProperty;
exports.unique = unique;
const _1 = __importDefault(require("."));
const util = __importStar(require("util"));
exports.validators = {
    required: _1.default.required,
    notEmptyString: _1.default.notEmptyString,
    rangeNumber: _1.default.ranges.number,
    rangeLength: _1.default.ranges.length,
    insideList: _1.default.lists.inside,
    outsideList: _1.default.lists.outside,
    password: _1.default.security.password,
    patterns: _1.default.patterns,
    equalToProperty: null, // Will be set below
    unique: null, // Will be set below
};
/**
 * Check if a value is the same as a value
 * of another property (useful for password
 * checking).
 **/
function equalToProperty(name, msg) {
    return function (v, next) {
        if (v === this[name]) {
            return next();
        }
        return next(msg || 'not-equal-to-property');
    };
}
/**
 * Check if a property is unique in the collection.
 * This can take a while because a query has to be made against the Model.
 *
 * Due to the async nature of node, and concurrent web server environments,
 * an index on the database column is the only way to guarantee uniqueness.
 *
 * For sensibility's sake, undefined and null values are ignored for uniqueness
 * checks.
 *
 * Options:
 *   ignoreCase: for postgres; mysql ignores case by default.
 *   scope: (Array) scope uniqueness to listed properties
 **/
function unique(...args) {
    let msg = null;
    let opts = {};
    for (const arg of args) {
        if (typeof arg === "string") {
            msg = arg;
        }
        else if (typeof arg === "object") {
            opts = arg;
        }
    }
    return function (v, next, ctx) {
        const self = this;
        const finish = (err) => {
            try {
                next(err);
            }
            catch (_invokeErr) {
                // Swallow downstream errors so validation pipeline isn't interrupted.
            }
        };
        (async () => {
            if (!ctx) {
                return finish();
            }
            if (typeof v === "undefined" || v === null) {
                return finish();
            }
            if (!ctx.model.properties[ctx.property]) {
                return finish();
            }
            // Cannot process on database engines which don't support SQL syntax
            if (!ctx.driver.isSql) {
                return finish('not-supported');
            }
            const chain = ctx.model.find();
            const chainQuery = (prop, value) => {
                let query = null;
                if (opts.ignoreCase === true && ctx.model.properties[prop] && ctx.model.properties[prop].type === 'text') {
                    query = util.format('LOWER(%s.%s) LIKE LOWER(?)', ctx.driver.query.escapeId(ctx.model.table), ctx.driver.query.escapeId(prop));
                    chain.where(query, [value]);
                }
                else {
                    query = {};
                    query[prop] = value;
                    chain.where(query);
                }
            };
            chainQuery(ctx.property, v);
            if (opts.scope) {
                for (const scopeProp of opts.scope) {
                    const scopedValue = ctx.instance[scopeProp];
                    if (typeof scopedValue === 'undefined' || scopedValue === null) {
                        return finish();
                    }
                    chainQuery(scopeProp, scopedValue);
                }
            }
            let records;
            try {
                records = await chain.run();
            }
            catch (err) {
                // Treat query errors as non-blocking to avoid cascading validation failures
                return finish();
            }
            if (!records || records.length === 0) {
                return finish();
            }
            const modelId = ctx.model.id;
            const idProperties = Array.isArray(modelId) ? modelId : [modelId];
            const otherRecords = records.filter((record) => {
                return !idProperties.every((prop) => record[prop] === self[prop]);
            });
            if (otherRecords.length === 0) {
                return finish();
            }
            return finish(msg || 'not-unique');
        })().catch(() => finish());
    };
}
// Add the ORM-specific validators to the validators object
exports.validators.equalToProperty = equalToProperty;
exports.validators.unique = unique;
exports.default = exports.validators;
