// orm3/dist/AggregateFunctions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
const ORMError = require("./Error");
const Utilities = __importStar(require("./Utilities"));
function AggregateFunctions(opts) {
    if (typeof opts.driver.getQuery !== "function") {
        throw new ORMError('NO_SUPPORT', "This driver does not support aggregate functions");
    }
    if (!Array.isArray(opts.driver.aggregate_functions)) {
        throw new ORMError('NO_SUPPORT', "This driver does not support aggregate functions");
    }
    var aggregates = [[]];
    var group_by = null;
    var used_distinct = false;
    var appendFunction = function (fun) {
        return function () {
            var args = (arguments.length && Array.isArray(arguments[0]) ? arguments[0] : Array.prototype.slice.apply(arguments));
            if (args.length > 0) {
                aggregates[aggregates.length - 1].push({ f: fun, a: args, alias: aggregateAlias(fun, args) });
                aggregates.push([]);
            }
            else {
                aggregates[aggregates.length - 1].push({ f: fun, alias: aggregateAlias(fun, args) });
            }
            if (fun === "distinct") {
                used_distinct = true;
            }
            return proto;
        };
    };
    var proto = {
        groupBy: function () {
            group_by = Array.prototype.slice.apply(arguments);
            return this;
        },
        limit: function (offset, limit) {
            if (typeof limit === "number") {
                opts.limit = [offset, limit];
            }
            else {
                opts.limit = [0, offset]; // offset = limit
            }
            return this;
        },
        order: function () {
            opts.order = Utilities.standardizeOrder(Array.prototype.slice.apply(arguments));
            return this;
        },
        select: function () {
            if (arguments.length === 0) {
                throw new ORMError('PARAM_MISMATCH', "When using append you must at least define one property");
            }
            if (Array.isArray(arguments[0])) {
                opts.propertyList = opts.propertyList.concat(arguments[0]);
            }
            else {
                opts.propertyList = opts.propertyList.concat(Array.prototype.slice.apply(arguments));
            }
            return this;
        },
        as: function (alias) {
            if (aggregates.length === 0 || (aggregates.length === 1 && aggregates[0].length === 0)) {
                throw new ORMError('PARAM_MISMATCH', "No aggregate functions defined yet");
            }
            var len = aggregates.length;
            aggregates[len - 1][aggregates[len - 1].length - 1].alias = alias;
            return this;
        },
        call: function (fun, args) {
            if (args && args.length > 0) {
                aggregates[aggregates.length - 1].push({ f: fun, a: args, alias: aggregateAlias(fun, args) });
                // aggregates.push([]);
            }
            else {
                aggregates[aggregates.length - 1].push({ f: fun, alias: aggregateAlias(fun, args) });
            }
            if (fun.toLowerCase() === "distinct") {
                used_distinct = true;
            }
            return this;
        },
        get: function (cb) {
            const execute = async () => {
                if (aggregates[aggregates.length - 1].length === 0) {
                    aggregates.length -= 1;
                }
                if (aggregates.length === 0) {
                    throw new ORMError('PARAM_MISMATCH', "Missing aggregate functions");
                }
                let query = opts.driver.getQuery().select().from(opts.table).select(opts.propertyList);
                for (let i = 0; i < aggregates.length; i++) {
                    for (let j = 0; j < aggregates[i].length; j++) {
                        query.fun(aggregates[i][j].f, aggregates[i][j].a, aggregates[i][j].alias);
                    }
                }
                query.where(opts.conditions);
                if (group_by !== null) {
                    query.groupBy.apply(query, group_by);
                }
                if (opts.order) {
                    for (let i = 0; i < opts.order.length; i++) {
                        query.order(opts.order[i][0], opts.order[i][1]);
                    }
                }
                if (opts.limit) {
                    query.offset(opts.limit[0]).limit(opts.limit[1]);
                }
                query = query.build();
                const data = await new Promise((resolve, reject) => {
                    opts.driver.execQuery(query, function (err, rows) {
                        if (err) {
                            reject(err);
                        }
                        else {
                            resolve(rows);
                        }
                    });
                });
                if (group_by !== null) {
                    return [data];
                }
                if (used_distinct && aggregates.length === 1) {
                    const columnKey = data.length > 0 ? Object.keys(data[0] || {}).pop() : undefined;
                    const items = data.map(item => (columnKey ? item[columnKey] : undefined));
                    return [items];
                }
                const items = [];
                const row = data[0] || {};
                for (let i = 0; i < aggregates.length; i++) {
                    for (let j = 0; j < aggregates[i].length; j++) {
                        items.push(row[aggregates[i][j].alias] ?? null);
                    }
                }
                return items;
            };
            if (typeof cb === 'function') {
                execute()
                    .then((values) => cb(null, ...values))
                    .catch((err) => cb(err));
                return;
            }
            return execute();
        }
    };
    for (var i = 0; i < opts.driver.aggregate_functions.length; i++) {
        addAggregate(proto, opts.driver.aggregate_functions[i], appendFunction);
    }
    return proto;
}
function addAggregate(proto, fun, builder) {
    if (Array.isArray(fun)) {
        proto[fun[0].toLowerCase()] = builder((fun[1] || fun[0]).toLowerCase());
    }
    else {
        proto[fun.toLowerCase()] = builder(fun.toLowerCase());
    }
}
function aggregateAlias(fun, fields) {
    return fun + (fields && fields.length ? "_" + fields.join("_") : "");
}
module.exports = AggregateFunctions;
