// orm3/dist/Error.js
"use strict";
/**
 * ORM Error Codes
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ORMError = exports.ErrorCodes = void 0;
exports.ErrorCodes = {
    QUERY_ERROR: 1,
    NOT_FOUND: 2,
    NOT_DEFINED: 3,
    NO_SUPPORT: 4,
    MISSING_CALLBACK: 5,
    PARAM_MISMATCH: 6,
    CONNECTION_LOST: 10,
    BAD_MODEL: 15,
};
/**
 * Custom ORM Error class
 */
class ORMError extends Error {
    constructor(message, code, extras) {
        super(message);
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
        this.name = 'ORMError';
        this.message = message;
        if (code) {
            this.code = exports.ErrorCodes[code];
            this.literalCode = code;
            if (!this.code) {
                throw new Error(`Invalid error code: ${code}`);
            }
        }
        if (extras) {
            Object.assign(this, extras);
        }
    }
    toString() {
        return `[ORMError ${this.literalCode}: ${this.message}]`;
    }
}
exports.ORMError = ORMError;
ORMError.codes = exports.ErrorCodes;
ORMError.ErrorCodes = exports.ErrorCodes;
exports.default = ORMError;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ORMError;
    module.exports.ORMError = ORMError;
    module.exports.ErrorCodes = exports.ErrorCodes;
    module.exports.codes = exports.ErrorCodes;
    module.exports.default = ORMError;
}
