// orm3/dist/LazyLoad.js
"use strict";
/**
 * Lazy loading utilities
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extend = extend;
function ucfirst(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}
function conditionAssign(instance, model) {
    const conditions = {};
    conditions[model.id] = instance[model.id];
    return conditions;
}
async function saveEntity(entity) {
    if (typeof entity.save !== 'function') {
        throw new Error('LazyLoad operations expect related instances to expose a save() method.');
    }
    if (entity.save.length > 0) {
        await new Promise((resolve, reject) => {
            entity.save((err) => {
                if (err) {
                    return reject(err);
                }
                resolve();
            });
        });
        return;
    }
    const result = entity.save();
    if (result && typeof result.then === 'function') {
        await result;
    }
}
async function fetchRelatedInstance(owner, Model, property) {
    const conditions = conditionAssign(owner, Model);
    const item = await Model
        .find(conditions, { identityCache: false })
        .only(Model.id.concat(property))
        .first();
    return item ?? null;
}
function addLazyLoadProperty(name, Instance, Model, property) {
    const method = ucfirst(name);
    const getterName = `get${method}`;
    const removerName = `remove${method}`;
    const setterName = `set${method}`;
    const getter = async function () {
        const item = await fetchRelatedInstance(this, Model, property);
        return item ? item[property] : null;
    };
    const remover = async function () {
        const item = await fetchRelatedInstance(this, Model, property);
        if (!item)
            return;
        item[property] = null;
        await saveEntity(item);
    };
    const setter = async function (data) {
        const item = await fetchRelatedInstance(this, Model, property);
        if (!item)
            return;
        item[property] = data;
        await saveEntity(item);
    };
    Object.defineProperty(Instance, getterName, {
        value: getter,
        enumerable: false
    });
    Object.defineProperty(Instance, setterName, {
        value: setter,
        enumerable: false
    });
    Object.defineProperty(Instance, removerName, {
        value: remover,
        enumerable: false
    });
}
function extend(Instance, Model, properties) {
    for (const k in properties) {
        if (properties[k].lazyload === true) {
            addLazyLoadProperty(properties[k].lazyname || k, Instance, Model, k);
        }
    }
}
exports.default = { extend };
