// orm3/dist/Settings.js
"use strict";
/**
 * Settings container for ORM configuration
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const _ = __importStar(require("lodash"));
const default_settings = {
    properties: {
        primary_key: "id",
        association_key: "{name}_{field}",
        required: false
    },
    instance: {
        identityCache: false,
        identityCacheSaveCheck: true,
        autoFetch: false,
        autoFetchLimit: 1,
        cascadeRemove: true,
        returnAllErrors: false,
        saveAssociationsByDefault: true
    },
    connection: {
        reconnect: true,
        pool: false,
        debug: false
    }
};
class SettingsContainerClass {
    constructor(settings) {
        this.settings = settings || _.cloneDeep(default_settings);
    }
    set(key, value) {
        setNested(key, value, this.settings);
        return this;
    }
    get(key, def) {
        let value = getNested(key, def, this.settings);
        if (typeof value === 'undefined' && typeof def === 'undefined' && key.indexOf('.') === -1) {
            for (const sectionKey of ['instance', 'properties', 'connection']) {
                const section = this.settings[sectionKey];
                if (section && Object.prototype.hasOwnProperty.call(section, key)) {
                    value = section[key];
                    break;
                }
            }
        }
        if (value instanceof Function) {
            return value;
        }
        else {
            return _.cloneDeep(value);
        }
    }
    unset(...keys) {
        for (const key of keys) {
            if (typeof key === "string") {
                unsetNested(key, this.settings);
            }
        }
        return this;
    }
}
const Settings = {
    Container: SettingsContainerClass,
    defaults: () => _.cloneDeep(default_settings)
};
function setNested(key, value, obj) {
    const p = key.indexOf(".");
    if (p === -1) {
        return obj[key] = value;
    }
    if (!obj.hasOwnProperty(key.substr(0, p))) {
        obj[key.substr(0, p)] = {};
    }
    return setNested(key.substr(p + 1), value, obj[key.substr(0, p)]);
}
function getNested(key, def, obj) {
    const p = key.indexOf(".");
    if (p === -1) {
        if (key === '*') {
            return obj;
        }
        return obj.hasOwnProperty(key) ? obj[key] : def;
    }
    if (!obj.hasOwnProperty(key.substr(0, p))) {
        return def;
    }
    return getNested(key.substr(p + 1), def, obj[key.substr(0, p)]);
}
function unsetNested(key, obj) {
    const p = key.indexOf(".");
    if (p === -1) {
        if (key === '*') {
            for (const k in obj) {
                delete obj[k];
            }
        }
        else {
            delete obj[key];
        }
        return;
    }
    if (!obj.hasOwnProperty(key.substr(0, p))) {
        return;
    }
    return unsetNested(key.substr(p + 1), obj[key.substr(0, p)]);
}
exports.default = Settings;
