// orm3/dist/Drivers/DDL/meta/postgresql/column.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgColumn = void 0;
const column_1 = require("../column");
class PgColumn extends column_1.Column {
    constructor(props) {
        super(props);
    }
    getName() {
        return this.meta.column_name;
    }
    isNullable() {
        return this.meta.is_nullable === 'YES';
    }
    getMaxLength() {
        return this.meta.character_maximum_length ?? null;
    }
    getDataType() {
        return this.meta.data_type;
    }
    isPrimaryKey() {
        return this.meta.is_primary_key === 'PRIMARY KEY';
    }
    getDefaultValue() {
        if (this.isAutoIncrementing()) {
            return null;
        }
        const defaultValue = this.meta.column_default;
        if (!defaultValue) {
            return null;
        }
        return defaultValue.replace(`::${this.meta.data_type}`, '');
    }
    isUnique() {
        return this.meta.column_key === 'UNIQUE' || this.meta.column_key === 'PRIMARY KEY';
    }
    isAutoIncrementing() {
        return (this.meta.column_default ?? '').startsWith('nextval');
    }
    getConstraintName() {
        return this.meta.constraint_name ?? null;
    }
    getReferencedTableName() {
        return this.meta.referenced_table_name ?? null;
    }
    getReferencedColumnName() {
        return this.meta.referenced_column_name ?? null;
    }
    getUpdateRule() {
        return this.meta.update_rule ?? null;
    }
    getDeleteRule() {
        return this.meta.delete_rule ?? null;
    }
}
exports.PgColumn = PgColumn;
