// orm3/dist/Drivers/DDL/sync/Queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
class Queue {
    constructor(cb) {
        this.cb = cb;
        this.pending = 0;
    }
    add(...args) {
        if (this.pending === -1) {
            return this;
        }
        this.pending += 1;
        const task = args.pop();
        if (typeof task !== "function") {
            throw new TypeError("Queue.add() requires a function as the last argument");
        }
        const wrappedCallback = (err) => {
            if (this.pending === -1) {
                return;
            }
            if (err) {
                this.pending = -1;
                this.cb(err);
                return;
            }
            if (--this.pending === 0) {
                this.cb();
            }
        };
        task(...args, wrappedCallback);
        return this;
    }
    check() {
        if (this.pending === 0) {
            this.cb();
        }
    }
}
exports.Queue = Queue;
exports.default = Queue;
