// orm3/dist/Drivers/DDL/sync/SQL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DROP_INDEX = exports.CREATE_INDEX = exports.ALTER_TABLE_DROP_COLUMN = exports.ALTER_TABLE_MODIFY_COLUMN = exports.ALTER_TABLE_RENAME_COLUMN = exports.ALTER_TABLE_ADD_COLUMN = exports.DROP_TABLE = exports.CREATE_TABLE = void 0;
const CREATE_TABLE = (options, driver) => {
    let sql = `CREATE TABLE ${driver.query.escapeId(options.name)} (${options.columns.join(", ")}`;
    if (options.keys && options.keys.length > 0) {
        const formattedKeys = options.keys.map((val) => driver.query.escapeId(val)).join(", ");
        sql += `, PRIMARY KEY (${formattedKeys})`;
    }
    sql += ")";
    return sql;
};
exports.CREATE_TABLE = CREATE_TABLE;
const DROP_TABLE = (options, driver) => {
    return `DROP TABLE ${driver.query.escapeId(options.name)}`;
};
exports.DROP_TABLE = DROP_TABLE;
const ALTER_TABLE_ADD_COLUMN = (options, driver) => {
    let sql = `ALTER TABLE ${driver.query.escapeId(options.name)} ADD ${options.column}`;
    if (options.after) {
        sql += ` AFTER ${driver.query.escapeId(options.after)}`;
    }
    else if (options.first) {
        sql += " FIRST";
    }
    return sql;
};
exports.ALTER_TABLE_ADD_COLUMN = ALTER_TABLE_ADD_COLUMN;
const ALTER_TABLE_RENAME_COLUMN = (options, driver) => {
    const eid = driver.query.escapeId;
    return `ALTER TABLE ${eid(options.name)} RENAME COLUMN ${eid(options.oldColName)} TO ${eid(options.newColName)}`;
};
exports.ALTER_TABLE_RENAME_COLUMN = ALTER_TABLE_RENAME_COLUMN;
const ALTER_TABLE_MODIFY_COLUMN = (options, driver) => {
    return `ALTER TABLE ${driver.query.escapeId(options.name)} MODIFY ${options.column}`;
};
exports.ALTER_TABLE_MODIFY_COLUMN = ALTER_TABLE_MODIFY_COLUMN;
const ALTER_TABLE_DROP_COLUMN = (options, driver) => {
    return `ALTER TABLE ${driver.query.escapeId(options.name)} DROP ${driver.query.escapeId(options.column)}`;
};
exports.ALTER_TABLE_DROP_COLUMN = ALTER_TABLE_DROP_COLUMN;
const CREATE_INDEX = (options, driver) => {
    const columns = options.columns.map((col) => driver.query.escapeId(col)).join(", ");
    return `CREATE${options.unique ? " UNIQUE" : ""} INDEX ${driver.query.escapeId(options.name)} ON ${driver.query.escapeId(options.collection)} (${columns})`;
};
exports.CREATE_INDEX = CREATE_INDEX;
const DROP_INDEX = (options, driver) => {
    return `DROP INDEX ${driver.query.escapeId(options.name)} ON ${driver.query.escapeId(options.collection)}`;
};
exports.DROP_INDEX = DROP_INDEX;
exports.default = {
    CREATE_TABLE: exports.CREATE_TABLE,
    DROP_TABLE: exports.DROP_TABLE,
    ALTER_TABLE_ADD_COLUMN: exports.ALTER_TABLE_ADD_COLUMN,
    ALTER_TABLE_RENAME_COLUMN: exports.ALTER_TABLE_RENAME_COLUMN,
    ALTER_TABLE_MODIFY_COLUMN: exports.ALTER_TABLE_MODIFY_COLUMN,
    ALTER_TABLE_DROP_COLUMN: exports.ALTER_TABLE_DROP_COLUMN,
    CREATE_INDEX: exports.CREATE_INDEX,
    DROP_INDEX: exports.DROP_INDEX,
};
