// orm3/dist/Drivers/DDL/meta/mysql/column.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MysqlColumn = void 0;
const column_1 = require("../column");
class MysqlColumn extends column_1.Column {
    constructor(props) {
        super(props);
    }
    getName() {
        return this.meta.column_name;
    }
    isNullable() {
        return this.meta.is_nullable === 'YES';
    }
    getMaxLength() {
        return this.meta.character_maximum_length ?? null;
    }
    getDataType() {
        return this.meta.data_type;
    }
    isPrimaryKey() {
        return this.meta.column_key === 'PRI';
    }
    getDefaultValue() {
        return this.meta.column_default;
    }
    isUnique() {
        return this.meta.column_key === 'UNI' || this.meta.column_key === 'PRI';
    }
    isAutoIncrementing() {
        return this.meta.extra === 'auto_increment';
    }
    getConstraintName() {
        return this.meta.constraint_name ?? null;
    }
    getReferencedTableName() {
        return this.meta.referenced_table_name ?? null;
    }
    getReferencedColumnName() {
        return this.meta.referenced_column_name ?? null;
    }
    getUpdateRule() {
        return this.meta.update_rule ?? null;
    }
    getDeleteRule() {
        return this.meta.delete_rule ?? null;
    }
}
exports.MysqlColumn = MysqlColumn;
