// orm3/dist/Drivers/DDL/sync/Dialects/questdb.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dropCollectionColumn = exports.modifyCollectionColumn = exports.addCollectionColumn = exports.dropCollection = exports.createCollection = exports.supportsType = exports.convertIndexes = exports.removeIndex = exports.addIndex = exports.getCollectionIndexes = exports.getCollectionProperties = exports.hasCollection = void 0;
const SQL = __importStar(require("../SQL"));
const postgresql_1 = __importDefault(require("./postgresql"));
const TYPE_MAP = {
    text: "STRING",
    integer: "INT",
    number: "DOUBLE",
    serial: "LONG",
    boolean: "BOOLEAN",
    date: "DATE",
    datetime: "TIMESTAMP",
    time: "TIMESTAMP",
    uuid: "UUID",
    point: "STRING",
    enum: "STRING",
    binary: "BLOB",
    object: "BLOB",
};
const normalizePropertyType = (property) => {
    if (!property) {
        return null;
    }
    if (property.__orm3PreferLong) {
        return TYPE_MAP.serial;
    }
    if (property.serial || property.type === "serial") {
        return TYPE_MAP.serial;
    }
    if (property.type === "date" && property.time) {
        return TYPE_MAP.datetime;
    }
    const mapped = TYPE_MAP[property.type];
    if (mapped) {
        return mapped;
    }
    if (property.rational === false && property.type === "number") {
        return TYPE_MAP.integer;
    }
    return "STRING";
};
const appendColumnQualifiers = (type, property, _driver) => {
    let result = type;
    if (property.required) {
        result += " NOT NULL";
    }
    // QuestDB currently rejects DEFAULT clauses in CREATE TABLE statements. The ORM
    // assigns default values at the application layer, so we purposely skip
    // emitting DEFAULT expressions during schema creation.
    return result;
};
const buildColumnDefinition = (property, driver) => {
    const mappedType = normalizePropertyType(property);
    if (!mappedType) {
        return false;
    }
    return appendColumnQualifiers(mappedType, property, driver);
};
const hasCollection = (driver, name, cb) => {
    const sql = `SELECT table_name FROM tables() WHERE table_name = ${driver.query.escapeVal(name)}`;
    driver.execSimpleQuery(sql, (err, rows) => {
        if (err) {
            cb(err);
            return;
        }
        cb(null, Array.isArray(rows) && rows.length > 0);
    });
};
exports.hasCollection = hasCollection;
const getCollectionProperties = (driver, name, cb) => {
    const sql = `SELECT column, type FROM table_columns(${driver.query.escapeVal(name)})`;
    driver.execSimpleQuery(sql, (err, rows) => {
        if (err) {
            cb(err);
            return;
        }
        const columns = {};
        for (const row of rows || []) {
            if (!row || !row.column) {
                continue;
            }
            columns[row.column] = {
                type: String(row.type || "").toLowerCase(),
            };
        }
        cb(null, columns);
    });
};
exports.getCollectionProperties = getCollectionProperties;
const getCollectionIndexes = (_driver, _name, cb) => {
    cb(null, {});
};
exports.getCollectionIndexes = getCollectionIndexes;
const addIndex = (_driver, _name, _unique, _collection, _columns, cb) => {
    cb(null);
};
exports.addIndex = addIndex;
const removeIndex = (_driver, _name, _collection, cb) => {
    cb(null);
};
exports.removeIndex = removeIndex;
const convertIndexes = (_collection, indexes) => indexes;
exports.convertIndexes = convertIndexes;
const supportsType = (type) => {
    switch (type) {
        case "serial":
            return TYPE_MAP.serial;
        case "integer":
            return TYPE_MAP.integer;
        case "number":
            return TYPE_MAP.number;
        case "boolean":
            return TYPE_MAP.boolean;
        case "date":
            return TYPE_MAP.date;
        case "enum":
            return TYPE_MAP.enum;
        case "uuid":
            return TYPE_MAP.uuid;
        case "text":
        default:
            return TYPE_MAP.text;
    }
};
exports.supportsType = supportsType;
const createCollection = (driver, name, columns, keys, cb) => {
    const sql = SQL.CREATE_TABLE({ name, columns, keys }, driver);
    driver.execSimpleQuery(sql, cb);
};
exports.createCollection = createCollection;
const dropCollection = (driver, name, cb) => {
    const sql = `DROP TABLE IF EXISTS ${driver.query.escapeId(name)}`;
    driver.execSimpleQuery(sql, cb);
};
exports.dropCollection = dropCollection;
const addCollectionColumn = (_driver, _name, _column, _afterColumn, cb) => {
    cb(null);
};
exports.addCollectionColumn = addCollectionColumn;
const modifyCollectionColumn = (_driver, _name, _column, cb) => {
    cb(null);
};
exports.modifyCollectionColumn = modifyCollectionColumn;
const dropCollectionColumn = (_driver, _name, _column, cb) => {
    cb(null);
};
exports.dropCollectionColumn = dropCollectionColumn;
const questdbDialect = {
    ...postgresql_1.default,
    hasCollection: exports.hasCollection,
    getCollectionProperties: exports.getCollectionProperties,
    getCollectionIndexes: exports.getCollectionIndexes,
    addIndex: exports.addIndex,
    removeIndex: exports.removeIndex,
    convertIndexes: exports.convertIndexes,
    processKeys: () => [],
    getType: (_collection, property, driver) => {
        const definition = buildColumnDefinition(property, driver);
        return definition === false ? false : { value: definition };
    },
    supportsType: exports.supportsType,
    createCollection: exports.createCollection,
    dropCollection: exports.dropCollection,
    addCollectionColumn: exports.addCollectionColumn,
    modifyCollectionColumn: exports.modifyCollectionColumn,
    dropCollectionColumn: exports.dropCollectionColumn,
};
exports.default = questdbDialect;
