// orm3/dist/utils/typeGuards.js
"use strict";
/**
 * Type Guards and Runtime Type Checking
 *
 * Provides type guard functions for runtime type checking and validation.
 * These help ensure type safety at runtime boundaries (e.g., callbacks, external input).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isError = isError;
exports.isErrorOrNull = isErrorOrNull;
exports.isORMErrorInfo = isORMErrorInfo;
exports.isQueryConditions = isQueryConditions;
exports.isValidID = isValidID;
exports.isIdentifiableEntity = isIdentifiableEntity;
exports.isInstanceData = isInstanceData;
exports.isDriverResult = isDriverResult;
exports.isHookMap = isHookMap;
exports.isFunction = isFunction;
exports.isCallable = isCallable;
exports.isValidationError = isValidationError;
exports.isValidationErrorArray = isValidationErrorArray;
exports.isPropertyDefinition = isPropertyDefinition;
exports.isAssociationDefinition = isAssociationDefinition;
exports.isArray = isArray;
exports.isEmptyArray = isEmptyArray;
exports.isNonEmptyArray = isNonEmptyArray;
exports.isStringArray = isStringArray;
exports.isNumberArray = isNumberArray;
exports.isString = isString;
exports.isNumber = isNumber;
exports.isBoolean = isBoolean;
exports.isNull = isNull;
exports.isUndefined = isUndefined;
exports.isNullish = isNullish;
exports.isTruthy = isTruthy;
exports.isFalsy = isFalsy;
exports.isPlainObject = isPlainObject;
exports.isObjectWithKeys = isObjectWithKeys;
exports.hasProperty = hasProperty;
exports.isDate = isDate;
exports.isDateString = isDateString;
exports.isTimestamp = isTimestamp;
exports.isErrorCallback = isErrorCallback;
exports.isResultCallback = isResultCallback;
exports.isVoidCallback = isVoidCallback;
// ==================== Error Checking ====================
/**
 * Check if value is an Error instance
 */
function isError(value) {
    return value instanceof Error;
}
/**
 * Check if value is an Error or null
 */
function isErrorOrNull(value) {
    return value === null || isError(value);
}
/**
 * Check if object is ORM error info
 */
function isORMErrorInfo(obj) {
    return (typeof obj === 'object' &&
        obj !== null &&
        'code' in obj &&
        'message' in obj &&
        typeof obj.message === 'string');
}
// ==================== Query Checking ====================
/**
 * Check if value is valid query conditions
 */
function isQueryConditions(obj) {
    return typeof obj === 'object' && obj !== null;
}
/**
 * Check if value is a valid ID (string, number, or array of them)
 */
function isValidID(value) {
    if (typeof value === 'string' || typeof value === 'number') {
        return true;
    }
    if (Array.isArray(value)) {
        return value.every(v => typeof v === 'string' || typeof v === 'number');
    }
    return false;
}
/**
 * Check if object is identifiable entity
 */
function isIdentifiableEntity(obj) {
    return (typeof obj === 'object' &&
        obj !== null &&
        !Array.isArray(obj) &&
        !(obj instanceof Date));
}
/**
 * Check if object is instance data
 */
function isInstanceData(obj) {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}
/**
 * Check if object is driver result
 */
function isDriverResult(obj) {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}
// ==================== Hook Checking ====================
/**
 * Check if value is a valid hook map
 */
function isHookMap(obj) {
    if (typeof obj !== 'object' || obj === null) {
        return false;
    }
    const hookNames = [
        'beforeCreate',
        'afterCreate',
        'beforeSave',
        'afterSave',
        'beforeValidation',
        'beforeRemove',
        'afterRemove',
        'afterLoad',
        'afterAutoFetch',
    ];
    for (const key in obj) {
        if (!hookNames.includes(key)) {
            return false;
        }
        const value = obj[key];
        if (value !== undefined && typeof value !== 'function') {
            return false;
        }
    }
    return true;
}
/**
 * Check if value is a function
 */
function isFunction(value) {
    return typeof value === 'function';
}
/**
 * Check if value is a callable
 */
function isCallable(value) {
    return typeof value === 'function';
}
// ==================== Validation Checking ====================
/**
 * Check if object is validation error
 */
function isValidationError(obj) {
    return (typeof obj === 'object' &&
        obj !== null &&
        'field' in obj &&
        'message' in obj &&
        'rule' in obj &&
        typeof obj.field === 'string' &&
        typeof obj.message === 'string' &&
        typeof obj.rule === 'string');
}
/**
 * Check if object is validation error array
 */
function isValidationErrorArray(obj) {
    return (Array.isArray(obj) &&
        obj.every(item => isValidationError(item)));
}
// ==================== Property Checking ====================
/**
 * Check if object is property definition
 */
function isPropertyDefinition(obj) {
    if (typeof obj !== 'object' || obj === null) {
        return false;
    }
    const def = obj;
    // Type check if present
    if (def.type !== undefined && typeof def.type !== 'string') {
        return false;
    }
    // Key check if present
    if (def.key !== undefined && typeof def.key !== 'boolean') {
        return false;
    }
    // Required check if present
    if (def.required !== undefined && typeof def.required !== 'boolean') {
        return false;
    }
    // Size check if present
    if (def.size !== undefined && typeof def.size !== 'number') {
        return false;
    }
    return true;
}
// ==================== Association Checking ====================
/**
 * Check if object is association definition
 */
function isAssociationDefinition(obj) {
    return (typeof obj === 'object' &&
        obj !== null &&
        'name' in obj &&
        'type' in obj &&
        'model' in obj &&
        typeof obj.name === 'string' &&
        typeof obj.type === 'string');
}
// ==================== Array & Collection Checking ====================
/**
 * Check if value is array
 */
function isArray(value) {
    return Array.isArray(value);
}
/**
 * Check if array is empty
 */
function isEmptyArray(value) {
    return Array.isArray(value) && value.length === 0;
}
/**
 * Check if array has items
 */
function isNonEmptyArray(value) {
    return Array.isArray(value) && value.length > 0;
}
/**
 * Check if value is array of strings
 */
function isStringArray(value) {
    return Array.isArray(value) && value.every(item => typeof item === 'string');
}
/**
 * Check if value is array of numbers
 */
function isNumberArray(value) {
    return Array.isArray(value) && value.every(item => typeof item === 'number');
}
// ==================== Primitive Checking ====================
/**
 * Check if value is string
 */
function isString(value) {
    return typeof value === 'string';
}
/**
 * Check if value is number
 */
function isNumber(value) {
    return typeof value === 'number' && !isNaN(value);
}
/**
 * Check if value is boolean
 */
function isBoolean(value) {
    return typeof value === 'boolean';
}
/**
 * Check if value is null
 */
function isNull(value) {
    return value === null;
}
/**
 * Check if value is undefined
 */
function isUndefined(value) {
    return value === undefined;
}
/**
 * Check if value is null or undefined
 */
function isNullish(value) {
    return value === null || value === undefined;
}
/**
 * Check if value is truthy
 */
function isTruthy(value) {
    return Boolean(value);
}
/**
 * Check if value is falsy
 */
function isFalsy(value) {
    return !value;
}
// ==================== Object Checking ====================
/**
 * Check if value is plain object (not array, Date, etc.)
 */
function isPlainObject(value) {
    return (typeof value === 'object' &&
        value !== null &&
        !Array.isArray(value) &&
        !(value instanceof Date) &&
        !(value instanceof Error) &&
        !(value instanceof RegExp));
}
/**
 * Check if value is plain object with specific keys
 */
function isObjectWithKeys(value, keys) {
    if (!isPlainObject(value)) {
        return false;
    }
    return keys.every(key => key in value);
}
/**
 * Check if object has a property
 */
function hasProperty(obj, key) {
    return (typeof obj === 'object' &&
        obj !== null &&
        key in obj);
}
// ==================== Date Checking ====================
/**
 * Check if value is Date
 */
function isDate(value) {
    return value instanceof Date && !isNaN(value.getTime());
}
/**
 * Check if value is valid date string (ISO 8601)
 */
function isDateString(value) {
    if (typeof value !== 'string') {
        return false;
    }
    const date = new Date(value);
    return !isNaN(date.getTime());
}
/**
 * Check if value is timestamp (milliseconds)
 */
function isTimestamp(value) {
    return typeof value === 'number' && value > 0 && value < Date.now() * 2;
}
// ==================== Callback Checking ====================
/**
 * Check if value is error callback
 */
function isErrorCallback(value) {
    return typeof value === 'function';
}
/**
 * Check if value is result callback
 */
function isResultCallback(value) {
    return typeof value === 'function';
}
/**
 * Check if value is void callback
 */
function isVoidCallback(value) {
    return typeof value === 'function';
}
// ==================== Export all guards ====================
exports.default = {
    isError,
    isErrorOrNull,
    isORMErrorInfo,
    isQueryConditions,
    isValidID,
    isIdentifiableEntity,
    isInstanceData,
    isDriverResult,
    isHookMap,
    isFunction,
    isCallable,
    isValidationError,
    isValidationErrorArray,
    isPropertyDefinition,
    isAssociationDefinition,
    isArray,
    isEmptyArray,
    isNonEmptyArray,
    isStringArray,
    isNumberArray,
    isString,
    isNumber,
    isBoolean,
    isNull,
    isUndefined,
    isNullish,
    isTruthy,
    isFalsy,
    isPlainObject,
    isObjectWithKeys,
    hasProperty,
    isDate,
    isDateString,
    isTimestamp,
    isErrorCallback,
    isResultCallback,
    isVoidCallback,
};
