// orm3/dist/SQLQuery/Comparators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.comparatorFunctions = exports.not_in = exports.lte = exports.lt = exports.gte = exports.gt = exports.ne = exports.eq = exports.not_like = exports.like = exports.not_between = exports.between = void 0;
function createSpecialObject(obj, tag) {
    Object.defineProperty(obj, 'sql_comparator', {
        configurable: false,
        enumerable: false,
        value: () => tag
    });
    return obj;
}
const between = (from, to) => createSpecialObject({ from, to }, 'between');
exports.between = between;
const not_between = (from, to) => createSpecialObject({ from, to }, 'not_between');
exports.not_between = not_between;
const like = (expr) => createSpecialObject({ expr }, 'like');
exports.like = like;
const not_like = (expr) => createSpecialObject({ expr }, 'not_like');
exports.not_like = not_like;
const eq = (val) => createSpecialObject({ val }, 'eq');
exports.eq = eq;
const ne = (val) => createSpecialObject({ val }, 'ne');
exports.ne = ne;
const gt = (val) => createSpecialObject({ val }, 'gt');
exports.gt = gt;
const gte = (val) => createSpecialObject({ val }, 'gte');
exports.gte = gte;
const lt = (val) => createSpecialObject({ val }, 'lt');
exports.lt = lt;
const lte = (val) => createSpecialObject({ val }, 'lte');
exports.lte = lte;
const not_in = (val) => createSpecialObject({ val }, 'not_in');
exports.not_in = not_in;
exports.comparatorFunctions = {
    between: exports.between,
    not_between: exports.not_between,
    like: exports.like,
    not_like: exports.not_like,
    eq: exports.eq,
    ne: exports.ne,
    gt: exports.gt,
    gte: exports.gte,
    lt: exports.lt,
    lte: exports.lte,
    not_in: exports.not_in
};
exports.default = exports.comparatorFunctions;
