// orm3/dist/SQLQuery/Create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateQuery = CreateQuery;
function tmpl(template, args) {
    const { hasOwnProperty } = Object.prototype;
    let result = template;
    for (const key in args) {
        if (!hasOwnProperty.call(args, key)) {
            continue;
        }
        const token = `{{${key}}}`;
        result = result.split(token).join(args[key]);
    }
    return result;
}
function buildFieldsList(dialect, structure) {
    if (!structure) {
        return '';
    }
    const tpl = "'{{NAME}}' {{TYPE}}";
    const fields = [];
    const { hasOwnProperty } = Object.prototype;
    const dataTypes = dialect.DataTypes || {};
    for (const field in structure) {
        if (!hasOwnProperty.call(structure, field)) {
            continue;
        }
        const dataType = dataTypes[structure[field]];
        const resolvedType = typeof dataType === 'string' ? dataType : structure[field];
        fields.push(tmpl(tpl, {
            NAME: field,
            TYPE: resolvedType
        }));
    }
    return fields.join(',');
}
function CreateQuery(dialect) {
    let tableName = null;
    let structure = {};
    return {
        table(name) {
            tableName = name;
            return this;
        },
        field(name, type) {
            structure[name] = type;
            return this;
        },
        fields(fields) {
            if (!fields) {
                return structure;
            }
            structure = fields;
            return this;
        },
        build() {
            if (!tableName) {
                return '';
            }
            const template = "CREATE TABLE '{{TABLE_NAME}}'({{FIELDS_LIST}})";
            const fieldsList = buildFieldsList(dialect, structure);
            return tmpl(template, {
                TABLE_NAME: tableName,
                FIELDS_LIST: fieldsList
            });
        }
    };
}
