// orm3/dist/SQLQuery/Query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = exports.Text = exports.Comparators = void 0;
const Create_1 = require("./Create");
const Select_1 = require("./Select");
const Insert_1 = require("./Insert");
const Update_1 = require("./Update");
const Remove_1 = require("./Remove");
const Comparators_1 = __importDefault(require("./Comparators"));
const Helpers = __importStar(require("./Helpers"));
const factory = (opts) => {
    let options;
    if (typeof opts === 'string') {
        options = { dialect: opts };
    }
    else {
        options = opts || {};
    }
    const dialectModule = require(`./Dialects/${options.dialect || 'mysql'}`);
    const Dialect = (dialectModule && dialectModule.default) ? dialectModule.default : dialectModule;
    return {
        escape: (query, args) => Helpers.escapeQuery(Dialect, query, args),
        escapeId: (...identifiers) => Dialect.escapeId(...identifiers),
        escapeVal: (value, timezone) => Dialect.escapeVal(value, timezone),
        create: () => (0, Create_1.CreateQuery)(Dialect),
        select: () => (0, Select_1.SelectQuery)(Dialect, options),
        insert: () => (0, Insert_1.InsertQuery)(Dialect, options),
        update: () => (0, Update_1.UpdateQuery)(Dialect, options),
        remove: () => (0, Remove_1.RemoveQuery)(Dialect, options)
    };
};
exports.Comparators = Object.keys(Comparators_1.default);
exports.Text = buildQueryType('text');
const QueryFactory = Object.assign(factory, {
    Comparators: exports.Comparators,
    Text: exports.Text
});
for (const comparator of exports.Comparators) {
    QueryFactory[comparator] = Comparators_1.default[comparator];
}
QueryFactory.Query = QueryFactory;
exports.Query = QueryFactory;
function buildQueryType(type) {
    return function build(data) {
        const result = { data };
        Object.defineProperty(result, 'type', {
            value: function typeGetter() {
                return type;
            },
            enumerable: false
        });
        return result;
    };
}
exports.default = QueryFactory;
