// orm3/dist/SQLQuery/Remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveQuery = RemoveQuery;
const Where_1 = require("./Where");
function RemoveQuery(dialect, opts = {}) {
    const sql = {
        where: [],
        order: []
    };
    return {
        from(table) {
            sql.table = table;
            return this;
        },
        where(...conditions) {
            for (let i = 0; i < conditions.length; i += 1) {
                sql.where.push({
                    t: null,
                    w: conditions[i]
                });
            }
            return this;
        },
        build() {
            const query = [];
            if (dialect.limitAsTop && Object.prototype.hasOwnProperty.call(sql, 'limit')) {
                query.push(`DELETE TOP ${sql.limit} FROM`);
            }
            else {
                query.push('DELETE FROM');
            }
            query.push(dialect.escapeId(sql.table));
            const whereClause = (0, Where_1.build)(dialect, sql.where, opts);
            if (Array.isArray(whereClause)) {
                query.push(...whereClause);
            }
            else if (whereClause) {
                query.push(whereClause);
            }
            if (sql.order.length > 0) {
                const orderParts = [];
                for (let i = 0; i < sql.order.length; i += 1) {
                    const orderSpec = sql.order[i];
                    if (Array.isArray(orderSpec.c)) {
                        orderParts.push(`${dialect.escapeId(orderSpec.c[0], orderSpec.c[1])} ${orderSpec.d}`);
                    }
                    else {
                        orderParts.push(`${dialect.escapeId(orderSpec.c)} ${orderSpec.d}`);
                    }
                }
                if (orderParts.length > 0) {
                    query.push(`ORDER BY ${orderParts.join(', ')}`);
                }
            }
            if (!dialect.limitAsTop) {
                if (Object.prototype.hasOwnProperty.call(sql, 'limit')) {
                    if (Object.prototype.hasOwnProperty.call(sql, 'offset')) {
                        query.push(`LIMIT ${sql.limit} OFFSET ${sql.offset}`);
                    }
                    else {
                        query.push(`LIMIT ${sql.limit}`);
                    }
                }
                else if (Object.prototype.hasOwnProperty.call(sql, 'offset')) {
                    query.push(`OFFSET ${sql.offset}`);
                }
            }
            return query.join(' ');
        },
        offset(offset) {
            sql.offset = offset;
            return this;
        },
        limit(limit) {
            sql.limit = limit;
            return this;
        },
        order(column, dir) {
            sql.order.push({
                c: Array.isArray(column) ? [column[0], column[1]] : column,
                d: dir === 'Z' ? 'DESC' : 'ASC'
            });
            return this;
        }
    };
}
