// orm3/dist/SQLQuery/Update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateQuery = UpdateQuery;
const Set_1 = require("./Set");
const Where_1 = require("./Where");
function UpdateQuery(dialect, opts = {}) {
    const sql = {
        where: []
    };
    return {
        into(table) {
            sql.table = table;
            return this;
        },
        set(values) {
            sql.set = values;
            return this;
        },
        where(...conditions) {
            for (let i = 0; i < conditions.length; i += 1) {
                sql.where.push({
                    t: null,
                    w: conditions[i]
                });
            }
            return this;
        },
        build() {
            const query = [];
            query.push('UPDATE');
            query.push(dialect.escapeId(sql.table));
            const setClause = (0, Set_1.build)(dialect, sql.set || {}, opts);
            if (Array.isArray(setClause)) {
                query.push(...setClause);
            }
            else if (setClause) {
                query.push(setClause);
            }
            const whereClause = (0, Where_1.build)(dialect, sql.where, opts);
            if (Array.isArray(whereClause)) {
                query.push(...whereClause);
            }
            else if (whereClause) {
                query.push(whereClause);
            }
            return query.join(' ');
        }
    };
}
