// orm3/dist/Drivers/DDL/meta/sqlite/column.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqliteColumn = void 0;
const column_1 = require("../column");
class SqliteColumn extends column_1.Column {
    constructor(props) {
        super(props);
    }
    getName() {
        return this.meta.name;
    }
    isNullable() {
        return this.meta.notnull === 0;
    }
    getMaxLength() {
        const match = this.getDataType().match(/\((\d+)\)$/);
        if (!match) {
            return null;
        }
        return Number(match[1]);
    }
    getDataType() {
        return this.meta.type.toUpperCase();
    }
    isPrimaryKey() {
        return this.meta.pk === 1;
    }
    getDefaultValue() {
        return this.meta.dflt_value;
    }
    isUnique() {
        return Boolean(this.meta.unique) || this.meta.pk === 1;
    }
    isAutoIncrementing() {
        return Boolean(this.meta.auto_increment);
    }
    markUnique() {
        this.meta.unique = true;
    }
    markAutoIncrement() {
        this.meta.auto_increment = true;
    }
}
exports.SqliteColumn = SqliteColumn;
