// orm3/dist/Validators/enforcements/patterns.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.match = match;
exports.hexString = hexString;
exports.email = email;
exports.ipv4 = ipv4;
exports.ipv6 = ipv6;
exports.mac = mac;
exports.uuid3 = uuid3;
exports.uuid4 = uuid4;
const Validator_1 = __importDefault(require("../Validator"));
function normalizeArgs(args) {
    if (typeof args[0] === 'string') {
        const [pattern, maybeModifiers, maybeMessage] = args;
        if (args.length === 2) {
            const message = maybeModifiers ?? 'no-pattern-match';
            return { pattern: new RegExp(pattern, 'i'), message };
        }
        const modifiers = maybeModifiers ?? 'i';
        const message = maybeMessage ?? 'no-pattern-match';
        return { pattern: new RegExp(pattern, modifiers), message };
    }
    const [regexp, maybeMessage] = args;
    return { pattern: regexp, message: maybeMessage ?? 'no-pattern-match' };
}
function match(...args) {
    const { pattern, message } = normalizeArgs(args);
    return new Validator_1.default((value, next) => {
        if (typeof value === 'string' && pattern.test(value))
            return next();
        return next(message);
    });
}
function hexString(message = 'not-hex-string') {
    return match('^[a-f0-9]+$', 'i', message);
}
function email(message = 'not-valid-email') {
    return match('^[a-z0-9\._%\+\-]+@[a-z0-9\.\-]+\.[a-z]{2,6}$', 'i', message);
}
function ipv4(message = 'not-valid-ipv4') {
    const p1 = '([1-9]|1[0-9][0-9]?|2[0-4][0-9]|25[0-4])';
    const p2 = '([0-9]|1[0-9][0-9]?|2[0-4][0-9]|25[0-4])';
    return match(`^${[p1, p2, p2, p1].join('\\.')}$`, '', message);
}
function ipv6(message = 'not-valid-ipv6') {
    const p1 = /^([a-f0-9]{1,4}:){7}[a-f0-9]{1,4}$/i;
    const p2 = /^([a-f0-9]{1,4}:)*[a-f0-9]{1,4}$/i;
    return new Validator_1.default((value, next) => {
        if (typeof value !== 'string')
            return next(message);
        if (value === '::' || value === '::1')
            return next();
        if (p1.test(value))
            return next();
        if (value.indexOf('::') === -1)
            return next(message);
        const group = value.split('::');
        if (group.length !== 2)
            return next(message);
        if (!p2.test(group[0]) || !p2.test(group[1]))
            return next(message);
        return next();
    });
}
function mac(message = 'not-valid-mac') {
    const p = '[0-9a-f]{1,2}';
    const s = '[\\.:]';
    return match(`^${[p, p, p, p, p, p].join(s)}$`, 'i', message);
}
function uuid3(message = 'not-valid-uuid3') {
    return match('^[a-f0-9]{8}\-[a-f0-9]{4}\-3[a-f0-9]{3}\-[89ab][a-f0-9]{3}\-[a-f0-9]{12}$', 'i', message);
}
function uuid4(message = 'not-valid-uuid3') {
    return match('^[a-f0-9]{8}\-[a-f0-9]{4}\-4[a-f0-9]{3}\-[89ab][a-f0-9]{3}\-[a-f0-9]{12}$', 'i', message);
}
